File: labs/lab03/backend/models/message.go
package models

import (
	"errors"
	"time"
)

// Message represents a chat message
type Message struct {
	ID        int       `json:"id"`
	Username  string    `json:"username"`
	Content   string    `json:"content"`
	Timestamp time.Time `json:"timestamp"`
}

// CreateMessageRequest represents the request to create a new message
type CreateMessageRequest struct {
	Username string `json:"username"` // validation: required
	Content  string `json:"content"`  // validation: required
}

// UpdateMessageRequest represents the request to update a message
type UpdateMessageRequest struct {
	Content string `json:"content"` // validation: required
}

// HTTPStatusResponse represents the response for HTTP status code endpoint
type HTTPStatusResponse struct {
	StatusCode  int    `json:"status_code"`
	ImageURL    string `json:"image_url"`
	Description string `json:"description"`
}

// APIResponse represents a generic API response
type APIResponse struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
}

// NewMessage creates a new message with the current timestamp
func NewMessage(id int, username, content string) *Message {
	return &Message{
		ID:        id,
		Username:  username,
		Content:   content,
		Timestamp: time.Now(),
	}
}

// Validate checks if the create message request is valid
func (r *CreateMessageRequest) Validate() error {
	if r.Username == "" {
		return errors.New("username is required")
	}
	if r.Content == "" {
		return errors.New("content is required")
	}
	return nil
}

// Validate checks if the update message request is valid
func (r *UpdateMessageRequest) Validate() error {
	if r.Content == "" {
		return errors.New("content is required")
	}
	return nil
}

File: labs/lab03/frontend/lib/models/message.dart
class Message {
  final int id;
  final String username;
  final String content;
  final DateTime timestamp;

  const Message({
    required this.id,
    required this.username,
    required this.content,
    required this.timestamp
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'],
      username: json['username'],
      content: json['content'],
      timestamp: DateTime.parse(json['timestamp'])
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'username': username,
      'content': content,
      'timestamp': timestamp.toIso8601String()
    };
  }
}

class CreateMessageRequest {
  final String username;
  final String content;

  const CreateMessageRequest({required this.username, required this.content});

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'username': username,
      'content': content
    };
  }

  String? validate() {
    if (username.isEmpty) {
      return 'Username is required';
    }
    if (content.isEmpty) {
      return 'Content is required';
    }
    return null;
  }
}

class UpdateMessageRequest {
  final String content;

  const UpdateMessageRequest({required this.content});

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'content': content
    };
  }

  String? validate() {
    if (content.isEmpty) {
      return 'Content is required';
    }
    return null;
  }
}

class HTTPStatusResponse {
  final int statusCode;
  final String imageUrl;
  final String description;

  const HTTPStatusResponse({
    required this.statusCode,
    required this.imageUrl,
    required this.description
  });

  factory HTTPStatusResponse.fromJson(Map<String, dynamic> json) {
    return HTTPStatusResponse(
        statusCode: json['status_code'],
        imageUrl: json['image_url'],
        description: json['description']
    );
  }
}

class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? error;

  const ApiResponse({required this.success, this.data, this.error});

  factory ApiResponse.fromJson(Map<String, dynamic> json, T Function(Map<String, dynamic>)? fromJsonT) {
    return ApiResponse(
      success: json['success'],
      error: json['error'],
      data: json['data'] != null && fromJsonT != null
          ? fromJsonT(json['data'])
          : null,
    );
  }
}
